### PR TITLE
implement demo cw1155 contract

### PR DESCRIPTION
**Pull request by yihuang**
there are several more unit tests to be done, the overall implementation is good for review now.
There are two `FIXME` currently:
- [code](https://github.com/CosmWasm/cosmwasm-plus/pull/251/files#diff-7a4ab5d1010ab74822d59ab1b9c2de4d53f46b0d1b7c88c3ba5fdff7ba186ddcR90), we should add overflow error branch in `StdError`.
- [code](https://github.com/CosmWasm/cosmwasm-plus/pull/251/files#diff-7a4ab5d1010ab74822d59ab1b9c2de4d53f46b0d1b7c88c3ba5fdff7ba186ddcR1102), the list returned by query is out of order, I'm not sure if it's an issue of the test storage itself?

### CODE REVIEW

File: contracts/cw1155-base/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,1283 @@
...
+        } => execute_burn(env, from, token_id, value),
+        Cw1155HandleMsg::BatchBurn { from, batch } => execute_batch_burn(env, from, batch),
+        Cw1155HandleMsg::ApproveAll { operator, expires } => {
+            execute_approve_all(env, operator, expires)
+        }
+        Cw1155HandleMsg::RevokeAll { operator } => execute_revoke_all(env, operator),
+    }
+}
+
+fn checked_add(a: Uint128, b: Uint128) -> StdResult<Uint128> {
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:86]: Interesting. We define `sub` for `Uint128` to possibly return an error if it underflows (not enough balance). We don't do this for adding as we assumed this would never happen. 

We do however, [test that this extremely rare condition causes a panic](https://github.com/CosmWasm/cosmwasm/blob/main/packages/std/src/math.rs#L774-L780), which is enough to keep the contract secure. It won't provide a pretty error message to the user, but will block the attack vector. I would just use the normal `Uint128 + Uint128` approach

File: contracts/cw1155-base/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,1283 @@
...
+                mock_info(users[0].clone(), &[]),
+                Cw1155HandleMsg::ApproveAll {
+                    operator: user.clone(),
+                    expires: None,
+                },
+            )
+            .unwrap();
+        }
+
+        // // FIXME test failed, query result out of order
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:1102]: They are not "out of order", it is that the keys are stored in `CanonicalAddr`, which you cannot assume is the same ordering as `HumanAddr`. For bech32 the order sometimes is the same, sometimes different. For tests, we semi-randomize the mapping to ensure you do not depend on it.

If you hardcode some addresses, just figure out via a few test runs what the ordering is as CanonicalAddr and then assert that one. It will be stable, but it makes you be sure if you are comparing canonical or human addresses.

Discussion here: https://github.com/CosmWasm/cosmwasm/issues/552

**Review by ethanfrey — COMMENTED** [review]
I just got back from vacation and have lots to do before I can take a deeper look, but I did add comments on the 2 FIXMEs. These are subtle points about how `cosmwasm-std` is designed and actually intended.

File: contracts/cw1155-base/src/contract.rs (hunk 1 above)
Comment (yihuang) [reply contracts/cw1155-base/src/contract.rs]: I reverted back to simple `+` operator.
Opened issue in cosmwasm-std for the overflow handling: https://github.com/CosmWasm/cosmwasm/issues/850

File: contracts/cw1155-base/src/contract.rs (hunk 2 above)
Comment (yihuang) [reply contracts/cw1155-base/src/contract.rs]: thanks for clarification, FIXME is resolved now.

**Comment by yihuang** [comment]
> I just got back from vacation and have lots to do before I can take a deeper look, but I did add comments on the 2 FIXMEs. These are subtle points about how `cosmwasm-std` is designed and actually intended.

Thanks, two FIXMEs are resolved now.

**Comment by yihuang** [comment]
Add unit tests to expect panic when minting token amount overflow.

File: contracts/cw1155-base/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+[package]
+name = "cw1155-base"
+version = "0.6.0-alpha1"
```
Comment (ethanfrey) [inline contracts/cw1155-base/Cargo.toml:3]: I think I just bumped this up to `-alpha3` trying to get multitest working for another app I'm building.

Comment (ethanfrey) [reply contracts/cw1155-base/Cargo.toml:3]: See #257 and #258

File: contracts/cw1155-base/src/state.rs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use cosmwasm_std::{CanonicalAddr, Uint128};
+use cw1155::Expiration;
+use cw_storage_plus::{Item, Map};
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/state.rs:8]: Where is this used? In query response?

File: contracts/cw1155-base/src/state.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+pub struct Approval {
+    /// Account that can transfer/send the token
+    pub spender: CanonicalAddr,
+    /// When the Approval expires (maybe Expiration::never)
+    pub expires: Expiration,
+}
+/// Store the minter address who have permission to mint new tokens.
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/state.rs:15]: Nice job with the state, and good to comment each one what those &[u8] and &str mean :+1:

File: packages/cw1155/src/event.rs [hunk 6]
```diff
@@ -1,42 +1,54 @@
-use cosmwasm_std::{attr, Attribute, HumanAddr, Uint128};
+use cosmwasm_std::{HumanAddr, Response, Uint128};
 use cw0::Event;
 
-use crate::msg::TokenId;
-
+/// Tracks token transfer/mint/burn actions
```
Comment (ethanfrey) [inline packages/cw1155/src/event.rs:4]: Interesting that this is all references.
Definitely more efficient, but harder to get write. But nicely done.

File: contracts/cw1155-base/src/contract.rs [hunk 7]
```diff
@@ -0,0 +1,1239 @@
...
+use crate::state::{APPROVES, BALANCES, MINTER, TOKENS};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cw1155-base";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+const DEFAULT_LIMIT: u32 = 10;
+const MAX_LIMIT: u32 = 30;
+
+pub fn instantiate(
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:26]: :100:

File: contracts/cw1155-base/src/contract.rs [hunk 8]
```diff
@@ -0,0 +1,1239 @@
...
+    _info: MessageInfo,
+    msg: InitMsg,
+) -> StdResult<Response> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let minter = deps.api.canonical_address(&msg.minter)?;
+    MINTER.save(deps.storage, &minter)?;
+    Ok(Response::default())
+}
+
+/// To mitigate clippy::too_many_arguments warning
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:38]: Interesting idea, I never tried that, but makes sense here

File: contracts/cw1155-base/src/contract.rs [hunk 9]
```diff
@@ -0,0 +1,1239 @@
...
+fn check_can_approve(
+    deps: Deps,
+    env: &Env,
+    owner: &HumanAddr,
+    operator: &HumanAddr,
+) -> StdResult<bool> {
+    // owner can approve
+    let owner_raw = deps.api.canonical_address(owner)?;
+    let operator_raw = deps.api.canonical_address(operator)?;
+    if owner_raw == operator_raw {
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:132]: Important check.
Good logic here

File: contracts/cw1155-base/src/contract.rs [hunk 10]
```diff
@@ -0,0 +1,1239 @@
...
+    to: Option<&'a HumanAddr>,
+    token_id: &'a str,
+    amount: Uint128,
+) -> Result<TransferEvent<'a>, ContractError> {
+    if let Some(from) = from {
+        let from_raw = deps.api.canonical_address(from)?;
+        BALANCES.update(
+            deps.storage,
+            (from_raw.as_slice(), token_id),
+            |balance: Option<Uint128>| balance.unwrap_or_default() - amount,
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:101]: Hmmm... do you want `checked_add` here? 
Or are you waiting for the next cosmwasm-std release?

File: contracts/cw1155-base/src/contract.rs [hunk 11]
```diff
@@ -0,0 +1,1239 @@
...
+            execute_approve_all(env, operator, expires)
+        }
+        Cw1155ExecuteMsg::RevokeAll { operator } => execute_revoke_all(env, operator),
+    }
+}
+
+/// When from is None: mint new coins
+/// When to is None: burn coins
+/// When both are None: not token balance is changed, meaningless but valid
+fn execute_transfer_inner<'a>(
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:89]: Nice internal abstraction (after permission checks are done)

File: contracts/cw1155-base/src/contract.rs [hunk 12]
```diff
@@ -0,0 +1,1239 @@
...
+    amount: Uint128,
+    msg: Option<Binary>,
+) -> Result<Response, ContractError> {
+    let ExecuteEnv {
+        mut deps,
+        env,
+        info,
+    } = env;
+
+    guard(
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:165]: You find this easier to write than

```rust
if !check_can_approve(deps.as_ref(), &env, &from, &info.sender)? {
  return Err(ContractError::Unauthorized {});
}
```

?? It's fine either way.

File: contracts/cw1155-base/src/contract.rs [hunk 13]
```diff
@@ -0,0 +1,1239 @@
...
+        check_can_approve(deps.as_ref(), &env, &from, &info.sender)?,
+        ContractError::Unauthorized {},
+    )?;
+
+    let mut rsp = Response::default();
+
+    let event = execute_transfer_inner(&mut deps, Some(&from), Some(&to), &token_id, amount)?;
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:172]: This is very cool use of events

File: contracts/cw1155-base/src/contract.rs [hunk 14]
```diff
@@ -0,0 +1,1239 @@
...
+    event.add_attributes(&mut rsp);
+
+    rsp.messages = if let Some(msg) = msg {
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:175]: How about just setting this inside the `if` and dropping the else?
`rsp.messages = vec![]` is already set with `default` above.

File: contracts/cw1155-base/src/contract.rs [hunk 15]
```diff
@@ -0,0 +1,1239 @@
...
+    if sender != MINTER.load(deps.storage)? {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let mut rsp = Response::default();
+
+    let event = execute_transfer_inner(&mut deps, None, Some(&to), &token_id, amount)?;
+    event.add_attributes(&mut rsp);
+
+    rsp.messages = if let Some(msg) = msg {
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:210]: same as above

File: contracts/cw1155-base/src/contract.rs [hunk 16]
```diff
@@ -0,0 +1,1239 @@
...
+        }
+        .into_cosmos_msg(to)?]
+    } else {
+        vec![]
+    };
+
+    // insert if not exist
+    let key = TOKENS.key(&token_id);
+    if deps.storage.get(&key).is_none() {
+        key.save(deps.storage, &"".to_owned())?;
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:226]: This means creating new token with no metadata?

Nice trick with `deps.storage.get(&key).is_none()` to do a "has" query and avoid parsing.
Should we support something like `TOKENS.has(&token_id)?` directly in cw-storage-plus?

File: contracts/cw1155-base/src/contract.rs [hunk 17]
```diff
@@ -0,0 +1,1239 @@
...
+
+    guard(
+        check_can_approve(deps.as_ref(), &env, &from, &info.sender)?,
+        ContractError::Unauthorized {},
+    )?;
+
+    let mut rsp = Response::default();
+    for (token_id, amount) in batch.iter() {
+        let event = execute_transfer_inner(&mut deps, Some(&from), Some(&to), token_id, *amount)?;
+        event.add_attributes(&mut rsp);
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:277]: And here the events really shine. :1st_place_medal:

File: contracts/cw1155-base/src/contract.rs [hunk 18]
```diff
@@ -0,0 +1,1239 @@
...
+    }
+
+    let mut rsp = Response::default();
+
+    for (token_id, amount) in batch.iter() {
+        let event = execute_transfer_inner(&mut deps, None, Some(&to), &token_id, *amount)?;
+        event.add_attributes(&mut rsp);
+    }
+
+    for (token_id, _) in batch.iter() {
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:314]: You don't want to roll this in the same above for loop?

File: contracts/cw1155-base/src/contract.rs [hunk 19]
```diff
@@ -0,0 +1,1239 @@
...
+
+    let mut rsp = Response::default();
+    for (token_id, amount) in batch.into_iter() {
+        let event = execute_transfer_inner(&mut deps, Some(&from), None, &token_id, amount)?;
+        event.add_attributes(&mut rsp);
+    }
+    Ok(rsp)
+}
+
+pub fn execute_approve_all(
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:361]: Looks good (approve and revoke)

File: contracts/cw1155-base/src/contract.rs [hunk 20]
```diff
@@ -0,0 +1,1239 @@
...
+        sender: &info.sender,
+        operator: &operator,
+        approved: false,
+    }
+    .add_attributes(&mut rsp);
+    Ok(rsp)
+}
+
+pub fn query(deps: Deps, env: Env, msg: Cw1155QueryMsg) -> StdResult<Binary> {
+    match msg {
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:406]: It does make sense to inline these short functions. I never thought of it.

File: contracts/cw1155-base/src/contract.rs [hunk 21]
```diff
@@ -0,0 +1,1239 @@
...
+        Cw1155QueryMsg::IsApprovedForAll { owner, operator } => {
+            let approved = check_can_approve(deps, &env, &owner, &operator)?;
+            to_binary(&IsApprovedForAllResponse { approved })
+        }
+        Cw1155QueryMsg::ApprovedForAll {
+            owner,
+            include_expired,
+            start_after,
+            limit,
+        } => to_binary(&query_all_approvals(
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:435]: And clearly some deserve their own function

File: contracts/cw1155-base/src/contract.rs [hunk 22]
```diff
@@ -0,0 +1,1239 @@
...
+    include_expired: bool,
+    start_after: Option<HumanAddr>,
+    limit: Option<u32>,
+) -> StdResult<ApprovedForAllResponse> {
+    let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+    let start_canon = maybe_canonical(deps.api, start_after)?;
+    let start = start_canon.map(Bound::exclusive);
+
+    let owner_raw = deps.api.canonical_address(&owner)?;
+    let operators = APPROVES
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:478]: Nice work with the ranges (here and below)

File: contracts/cw1155-base/src/contract.rs [hunk 23]
```diff
@@ -0,0 +1,1239 @@
...
+                mock_info(user1.clone(), &[]),
+                Cw1155ExecuteMsg::BatchBurn {
+                    from: user1.clone(),
+                    batch: vec![(token2.clone(), 1u64.into()), (token3.clone(), 1u64.into())]
+                }
+            )
+            .unwrap(),
+            Response {
+                attributes: vec![
+                    attr("action", "transfer"),
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:881]: This is interesting. I am not sure how this ends up when it comes out through tendermint (I think it does some "collapsing" of events).

It totally makes sense. I wonder if you had a wish on a cleaner way to encode these or pass them through to the client? Although I guess just as it is is perfect for searching.

File: contracts/cw1155-base/src/state.rs (hunk 4 above)
Comment (yihuang) [reply contracts/cw1155-base/src/state.rs:8]: it's initially copied from cw721 and never used, removed now.

File: contracts/cw1155-base/Cargo.toml (hunk 3 above)
Comment (yihuang) [reply contracts/cw1155-base/Cargo.toml:3]: fixed now.

File: contracts/cw1155-base/src/contract.rs (hunk 10 above)
Comment (yihuang) [reply contracts/cw1155-base/src/contract.rs:101]: I think we can change them at once in the cosmwasm-std upgrade.

File: contracts/cw1155-base/src/contract.rs (hunk 12 above)
Comment (yihuang) [reply contracts/cw1155-base/src/contract.rs:170]: agree, it's pointless, removed the `guard`, ;D

File: contracts/cw1155-base/src/contract.rs (hunk 14 above)
Comment (yihuang) [reply contracts/cw1155-base/src/contract.rs:180]: done

File: contracts/cw1155-base/src/contract.rs (hunk 15 above)
Comment (yihuang) [reply contracts/cw1155-base/src/contract.rs:215]: done

File: contracts/cw1155-base/src/contract.rs (hunk 16 above)
Comment (yihuang) [reply contracts/cw1155-base/src/contract.rs:231]: > This means creating new token with no metadata?

yes, insert an empty entry so token enumeration can find it.

> 
> Nice trick with `deps.storage.get(&key).is_none()` to do a "has" query and avoid parsing.
> Should we support something like `TOKENS.has(&token_id)?` directly in cw-storage-plus?

I think that'll be useful, another option is `key.exists(deps.storage)`.

File: contracts/cw1155-base/src/contract.rs (hunk 18 above)
Comment (yihuang) [reply contracts/cw1155-base/src/contract.rs:319]: > You don't want to roll this in the same above for loop?

thanks, rolled in now, looks better.

File: contracts/cw1155-base/src/contract.rs (hunk 23 above)
Comment (yihuang) [reply contracts/cw1155-base/src/contract.rs:866]: I'm not sure how it's going to be indexed by tendermint though, need to look into details in tendermint I guess.

File: contracts/cw1155-base/src/contract.rs (hunk 12 above)
Comment (yihuang) [reply contracts/cw1155-base/src/contract.rs]: Further encapsulate a `guard_can_approve` which return `ContractError` directly.

Comment (ethanfrey) [reply contracts/cw1155-base/src/contract.rs]: > Further encapsulate a guard_can_approve which return ContractError directly.

Yes, something like `guard_can_approve(deps.as_ref(), &env, &from, &info.sender)?;` or `assert_can_approve(deps.as_ref(), &env, &from, &info.sender)?;` seems like a nice abstraction, just not the two separately which doesn't save typing

File: contracts/cw1155-base/src/contract.rs [hunk 24]
```diff
@@ -207,18 +201,16 @@ pub fn execute_mint(
     let event = execute_transfer_inner(&mut deps, None, Some(&to), &token_id, amount)?;
     event.add_attributes(&mut rsp);
 
-    rsp.messages = if let Some(msg) = msg {
-        vec![Cw1155ReceiveMsg {
+    if let Some(msg) = msg {
+        rsp.messages = vec![Cw1155ReceiveMsg {
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:216]: This could also be `rsp.add_message(Cw1155ReceiveMsg { .. }.into_cosmos_msg(to)?);`

But I don't think that is any clearer in this case.

File: contracts/cw1155-base/src/contract.rs [hunk 25]
```diff
@@ -309,26 +297,23 @@ pub fn execute_batch_mint(
     for (token_id, amount) in batch.iter() {
         let event = execute_transfer_inner(&mut deps, None, Some(&to), &token_id, *amount)?;
         event.add_attributes(&mut rsp);
-    }
 
-    for (token_id, _) in batch.iter() {
         // insert if not exist
         let key = TOKENS.key(&token_id);
         if deps.storage.get(&key).is_none() {
+            // insert an empty entry so token enumeration can find it
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:311]: thanks for the comment

File: contracts/cw1155-base/src/contract.rs [hunk 26]
```diff
@@ -42,6 +44,7 @@ pub struct ExecuteEnv<'a> {
     info: MessageInfo,
 }
 
+#[cfg_attr(not(feature = "library"), entry_point)]
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:47]: Nice update

File: contracts/cw1155-base/src/contract.rs [hunk 27]
```diff
@@ -159,9 +172,7 @@ pub fn execute_send_from(
         info,
     } = env;
 
-    if !check_can_approve(deps.as_ref(), &env, &from, &info.sender)? {
-        return Err(ContractError::Unauthorized {});
-    }
+    guard_can_approve(deps.as_ref(), &env, &from, &info.sender)?;
```
Comment (ethanfrey) [inline contracts/cw1155-base/src/contract.rs:175]: Yeah, this is a nicer way to inline the common check.

File: contracts/cw1155-base/src/contract.rs (hunk 23 above)
Comment (ethanfrey) [reply contracts/cw1155-base/src/contract.rs:871]: Yeah, they should all match

**Review by ethanfrey — APPROVED** [review]
Good stuff.

**Comment by ethanfrey** [comment]
Created https://github.com/CosmWasm/cosmwasm-plus/issues/261 inspired by your comments
